psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -0,0 +1,211 @@
+ * Copyright 2013-2021 The Regents of the University of California
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.linqs.psl.evaluation.statistics;
+
+import org.linqs.psl.application.learning.weight.TrainingMap;
+import org.linqs.psl.config.Options;
+import org.linqs.psl.model.atom.GroundAtom;
+import org.linqs.psl.model.atom.ObservedAtom;
+import org.linqs.psl.model.atom.RandomVariableAtom;
+import org.linqs.psl.model.predicate.Predicate;
+import org.linqs.psl.model.predicate.StandardPredicate;
+import org.linqs.psl.model.term.Constant;
+import org.linqs.psl.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Compute ranking-based statistics.
+ */
+public class RankingEvaluator extends Evaluator {
+    public enum RepresentativeMetric {
+        MRR
+    }
+
+    public static final String DELIM = ":";
+
+    private RepresentativeMetric representative;
+    private String defaultPredicate;
+    private List<Integer> categoryIndexes;
+
+    // This is the truth value threshold that truth atoms must pass
+    // for us to use their corresponding target atom in a ranking
+    // metric calculation.
+    private double threshold;
+
+    // Save the map from target atoms to truth atoms
+    // for truth value lookups in ranking metric calculations.
+    private Map<RandomVariableAtom, ObservedAtom> labelMap;
+
+     // Maintain a map from predicted atoms
+     // to their ranks (with respect to their individual
+     // entities) for use in ranking metric calculations.
+    private Map<GroundAtom, Integer> rankMap;
+
+    public RankingEvaluator() {
+        this(RepresentativeMetric.valueOf(Options.EVAL_RANK_REPRESENTATIVE.getString()),
+                StringUtils.splitInt(Options.EVAL_RANK_CATEGORY_INDEXES.getString(), DELIM));
+    }
+
+    public RankingEvaluator(int... rawCategoryIndexes) {
+        this(Options.EVAL_RANK_REPRESENTATIVE.getString(), rawCategoryIndexes);
+    }
+
+    public RankingEvaluator(String representative, int... rawCategoryIndexes) {
+        this(RepresentativeMetric.valueOf(representative.toUpperCase()), rawCategoryIndexes);
+    }
+
+    public RankingEvaluator(RepresentativeMetric representative, int... rawCategoryIndexes) {
+        this.representative = representative;
+
+        categoryIndexes = new ArrayList<Integer>(rawCategoryIndexes.length);
+
```

Comment:
Remove newline.
We want readers to see creating and populating the list as a single action (since population is trivial).